### PR TITLE
feat: rush: added env variables support. Created exit, echo, pwd builtins

### PR DESCRIPTION
**Pull request by xplshn**
Added env variables support and exit, echo, pwd builtins to `rush`.

###### hope this pr is not _ignored_ just like the others...              ;(

### CODE REVIEW

**Review by rminnich — APPROVED** [review]
fwiw, cd and info were maintained in separate files to show people how to add builtins, but I doubt this matters.

File: cmds/exp/rush/builtins.go [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+	"strconv"
+	"strings"
+)
+
+// Error messages
+var errCdUsage = errors.New("usage: cd <directory-to-change-to>")
+var errExitUsage = errors.New("usage: exit <|0-255|>")
+
+// Add builtins to the shell
+func init() {
```
Comment (rminnich) [inline cmds/exp/rush/builtins.go:24]: This is actually not really the best way to do this IMHO. If you look at the old cd.go, you can see that its init function did what you are doing here. Nobody cares about rush any more, so I'm not going to worry about it, but I actually think this is taking things the wrong direction.

**Review by rminnich — APPROVED** [review]

**Comment by rminnich** [comment]
You are missing a DCO. It's not quite fair to complain about reviews when you miss something as important as the DCO.

**Review by rminnich — CHANGES_REQUESTED** [review]
I'm going to withdraw my approval after all. The added init function in builtins.go is a mistake, and in many ways, the whole file is, because it misses the point that each builtin can be self-contained in its own file.

Rust was designed to be trivially extended just by dropping more files into this directory; in the original u-root it was compiled on demand. I don't see this change as an improvement. Please undo the changes to cd.go and info.go.

Thanks for the environment stuff. I preferred having /env, as in plan 9, but that idea is too foreign to unix, so your work is an improvement.

**Comment by xplshn** [comment]
Should I create internals.go? How should I name the file containing exit, pwd and echo? Or should each be a separate file?

**Comment by xplshn** [comment]
> You are missing a DCO. It's not quite fair to complain about reviews when you miss something as important as the DCO.

Sorry, I don't usually use the `-s` flag. I did `git commit --amend --no-edit --signoff`, which should set it now. (iThink). I'll keep it in mind for future contributions/commits.

**Comment by rminnich** [comment]
just name the files as they were:
cd.go
exit.go
and for each command you add, you can just add a file. The reason to do this is that (at least to me) it's clearer when the builtin is in its own file. Some builtins can get very large. 

The init function in those files adds the command to builtins. I realize this can result in lots of fiddly little files, and we can re-evaluate, but I think it's easier for newcomers to see how to easily add a new builtin.

Also, as to the command parser, if you want to make that better, go for it :-) The Plan 9 rc shell has a pretty interesting model for parsing, might be worth a look.

thanks

**Comment by xplshn** [comment]
> just name the files as they were: cd.go exit.go and for each command you add, you can just add a file. The reason to do this is that (at least to me) it's clearer when the builtin is in its own file. Some builtins can get very large.
> 
> The init function in those files adds the command to builtins. I realize this can result in lots of fiddly little files, and we can re-evaluate, but I think it's easier for newcomers to see how to easily add a new builtin.
> 
> Also, as to the command parser, if you want to make that better, go for it :-) The Plan 9 rc shell has a pretty interesting model for parsing, might be worth a look.
> 
> thanks

I came across an RC shell implementation in Go that I really liked. If POSIX compliance isn't too strict, we could incorporate support for lists and other features that are genuinely cool and useful. POSIX is primarily a compilation of past mistakes that were popularized by commercial Unix systems and their variations.

I'm a big fan of Plan9/9front, but I also really love Unix, I'd preffer things to be either entirely POSIX or completely new/different. I don't like blurry lines.

**Comment by rminnich** [comment]
I agree about not linking blurry lines, but the lines started blurring in the 1980s. It's not  a problem I want to solve.

I'm very interested in the rc shell, do you have a link?

**Comment by xplshn** [comment]
> I agree about not linking blurry lines, but the lines started blurring in the 1980s. It's not a problem I want to solve.
> 
> I'm very interested in the rc shell, do you have a link?

I just got back from a short trip, sorry for the delayed response, here's the link; https://github.com/as/rc

**Comment by xplshn** [comment]
Is this going to get merged as it is now? Are any changes needed?

**Comment by rminnich** [comment]
I don't think it should be merged as it is.

**Review by rminnich — COMMENTED** [review]
if you can get CI to pass we'll look again.

**Pull request closed without merging**